bigint/src/biguint_tests.rs

```diff
@@ -0,0 +1,1236 @@
+            Err(_) => panic!(),
+        };
+        assert_eq!(n, ans);
+    }
+
+    check(3, "6");
+    check(10, "3628800");
+    check(20, "2432902008176640000");
+    check(30, "265252859812191058636308480000000");
+}
+
+#[test]
+fn test_bits() {
+    assert_eq!(BigUint::new(vec![0, 0, 0, 0]).bits(), 0);
+    let n: BigUint = FromPrimitive::from_usize(0).unwrap();
+    assert_eq!(n.bits(), 0);
+    let n: BigUint = FromPrimitive::from_usize(1).unwrap();
+    assert_eq!(n.bits(), 1);
+    let n: BigUint = FromPrimitive::from_usize(3).unwrap();
+    assert_eq!(n.bits(), 2);
+    let n: BigUint = BigUint::from_str_radix("4000000000", 16).unwrap();
+    assert_eq!(n.bits(), 39);
+    let one: BigUint = One::one();
+    assert_eq!((one << 426).bits(), 427);
+}
+
+#[test]
+fn test_rand() {
+    let mut rng = thread_rng();
+    let _n: BigUint = rng.gen_biguint(137);
+    assert!(rng.gen_biguint(0).is_zero());
+}
+
+#[test]
+fn test_rand_range() {
+    let mut rng = thread_rng();
+
+    for _ in 0..10 {
+        assert_eq!(rng.gen_bigint_range(&FromPrimitive::from_usize(236).unwrap(),
+                                        &FromPrimitive::from_usize(237).unwrap()),
+                   FromPrimitive::from_usize(236).unwrap());
+    }
+
+    let l = FromPrimitive::from_usize(403469000 + 2352).unwrap();
+    let u = FromPrimitive::from_usize(403469000 + 3513).unwrap();
+    for _ in 0..1000 {
+        let n: BigUint = rng.gen_biguint_below(&u);
+        assert!(n < u);
+
+        let n: BigUint = rng.gen_biguint_range(&l, &u);
+        assert!(n >= l);
+        assert!(n < u);
+    }
+}
+
+#[test]
+#[should_panic]
+fn test_zero_rand_range() {
+    thread_rng().gen_biguint_range(&FromPrimitive::from_usize(54).unwrap(),
+                                   &FromPrimitive::from_usize(54).unwrap());
+}
+
+#[test]
+#[should_panic]
+fn test_negative_rand_range() {
+    let mut rng = thread_rng();
+    let l = FromPrimitive::from_usize(2352).unwrap();
+    let u = FromPrimitive::from_usize(3513).unwrap();
+    // Switching u and l should fail:
+    let _n: BigUint = rng.gen_biguint_range(&u, &l);
+}
+
+fn test_mul_divide_torture_count(count: usize) {
+    use rand::{SeedableRng, StdRng, Rng};
+
+    let bits_max = 1 << 12;
+    let seed: &[_] = &[1, 2, 3, 4];
+    let mut rng: StdRng = SeedableRng::from_seed(seed);
+
+    for _ in 0..count {
+        // Test with numbers of random sizes:
+        let xbits = rng.gen_range(0, bits_max);
+        let ybits = rng.gen_range(0, bits_max);
+
+        let x = rng.gen_biguint(xbits);
+        let y = rng.gen_biguint(ybits);
+
+        if x.is_zero() || y.is_zero() {
+            continue;
+        }
+
+        let prod = &x * &y;
+        assert_eq!(&prod / &x, y);
+        assert_eq!(&prod / &y, x);
+    }
+}
+
+#[test]
+fn test_mul_divide_torture() {
+    test_mul_divide_torture_count(1000);
+}
+
+#[test]
+#[ignore]
+fn test_mul_divide_torture_long() {
+    test_mul_divide_torture_count(1000000);
+}
```

Comment:
I suppose I can do that.

On Fri, Jul 15, 2016 at 1:42 AM, Łukasz Jan Niemier <
notifications@github.com> wrote:

> In bigint/src/biguint_tests.rs
> <https://github.com/rust-num/num/pull/207#discussion_r70947125>:
>
> > +        let prod = &x * &y;
> > +        assert_eq!(&prod / &x, y);
> > +        assert_eq!(&prod / &y, x);
> > +    }
> > +}
> > +
> > +#[test]
> > +fn test_mul_divide_torture() {
> > +    test_mul_divide_torture_count(1000);
> > +}
> > +
> > +#[test]
> > +#[ignore]
> > +fn test_mul_divide_torture_long() {
> > +    test_mul_divide_torture_count(1000000);
> > +}
>
> This file should be moved to test directory.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-num/num/pull/207/files/fb286a12a4c07883ab3a26def303e25a26ea0736#r70947125>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB5r7mNzVMYuWrMz2kVTleKmIAhG2d3Dks5qV1X8gaJpZM4JNNnq>
> .
>